graphql/src/store/prefetch.rs

```diff
@@ -321,92 +370,43 @@ impl<'a> Join<'a> {
             return;
         }
 
-        // Organize the join conditions by child type. We do not precompute that
-        // in `new`, because `perform` is only called once for each instance
-        // of a `Join`. For each child type, there might be multiple parent
-        // types that require different ways of joining to them
-        let conds_by_child: HashMap<_, Vec<_>> =
-            self.conds.iter().fold(HashMap::default(), |mut map, cond| {
-                map.entry(cond.child_type).or_default().push(cond);
-                map
-            });
-
-        // Build a map (parent_type, child_key) -> Vec<child> for joining by grouping
-        // children by their child_field
-        let mut grouped: BTreeMap<(&str, &str), Vec<Rc<Node>>> = BTreeMap::default();
+        // Build a map parent_id -> Vec<child> that we will use to add
+        // children to their parent. This relies on the fact that interfaces
+        // make sure that id's are distinct across all implementations of the
+        // interface.
+        let mut grouped: BTreeMap<&str, Vec<Rc<Node>>> = BTreeMap::default();
         for child in children.iter() {
-            for cond in conds_by_child.get(child.typename()).expect(&format!(
-                "query results only contain known types: {}",
-                child.typename()
-            )) {
-                match child
-                    .get("g$parent_id")
-                    .expect("the query that produces 'child' ensures there is always a g$parent_id")
-                {
-                    StoreValue::String(key) => grouped
-                        .entry((cond.parent_type, key))
-                        .or_default()
-                        .push(child.clone()),
-                    StoreValue::List(list) => {
-                        for key in list {
-                            match key {
-                                StoreValue::String(key) => grouped
-                                    .entry((cond.parent_type, key))
-                                    .or_default()
-                                    .push(child.clone()),
-                                _ => unreachable!("a list of join keys contains only strings"),
-                            }
-                        }
-                    }
-                    _ => unreachable!("join fields are strings or lists"),
-                }
+            match child
+                .get("g$parent_id")
+                .expect("the query that produces 'child' ensures there is always a g$parent_id")
+            {
+                StoreValue::String(key) => grouped.entry(key).or_default().push(child.clone()),
```

Comment:
I looked at that too, but didn't find a way to avoid the clone - everything I tried ended in borrow checker unhappiness. But child is an `Rc<Node>` so the clone is pretty cheap.

Reply:
Cloning an `Rc` is basically as cheap as addition. Some advocate using `Rc::clone(&child)` to make explicit that this clone is cheap.

Reply:
Oh, sorry, I thought it was a plain `Node`.